Google H5 Games Ads tutorial

yaustar opened the pull request:
YouTube link: https://www.youtube.com/watch?v=K2WWSnfJk4o&feature=emb_imp_woyt (Watermark and 720p resolution due to using the free version of Descript).

I'm not that convinced by the video tutorial at the moment and may remove it from this tutorial pending on feedback on the PR.

![Screenshot 2022-03-28 at 18-04-02 Google Ads to monetize your Game Learn PlayCanvas](https://user-images.githubusercontent.com/16639049/160512072-5863702e-2ff8-4e20-b7dc-a9a1282abfcf.png)

yaustar commented:
🤦 I've forgotten the step where you need to signup to adsense

willeastcott commented:
Please delete `content/_tutorial_contents.json` and `content/_tutorial_contents_legacy.json` from your PR (and your local file system).

willeastcott commented on content/en/tutorials/google-ads-for-games.md:
```diff
@@ -0,0 +1,377 @@
+---
+title: Google Ads to monetize your Game
```
```suggestion
title: Google Ads to Monetize your Game
```

willeastcott commented on content/en/tutorials/google-ads-for-games.md:
```diff
@@ -0,0 +1,377 @@
+
+## Introduction
+
+This tutorial will show you how to integrate Ads from Google AdSense into your games and help monetize.
```
```suggestion
This tutorial will show you how to integrate Ads from Google AdSense into your games for the purpose of monetization.
```

willeastcott commented on content/en/tutorials/google-ads-for-games.md:
```diff
@@ -0,0 +1,377 @@
+
+The Ads come in two main forms:
+
+Interstitial Ads that can appear between sessions of play such as going from game over back to the main menu (think of them as the ad breaks on TV shows)
```
```suggestion
Interstitial Ads that can appear between sessions of play such as going from 'game over' back to the main menu. Think of them as the ad breaks on TV shows.
```

willeastcott commented on content/en/tutorials/google-ads-for-games.md:
```diff
@@ -0,0 +1,377 @@
+
+![][flappy-bird-demo-rewarded-ad]
+
+For the tutorial we are going to fork the [Google H5 Ad Tutorial (Start)][tutorial-template-start] and add the ad SDK library and functionality  bit by bit.
```
```suggestion
For the tutorial we are going to fork the [Google H5 Ad Tutorial (Start)][tutorial-template-start] project and add the ad SDK library and functionality bit by bit.
```

willeastcott commented on content/en/tutorials/google-ads-for-games.md:
```diff
@@ -0,0 +1,377 @@
+
+Go to the [Google H5 Ad Tutorial (Start)][tutorial-template-start] project dashboard page and click on 'fork' and let's take a look at the project.
+
+This is a simple project where it's ready for us to add the button click callbacks to show the ads in the `ui-controller.js`
```
```suggestion
This is a simple project where it's ready for us to add the button click callbacks to show the ads in the `ui-controller.js` script:
```

willeastcott commented on content/en/tutorials/google-ads-for-games.md:
````diff
@@ -0,0 +1,377 @@
+};
+```
+
+First step is to add the Google SDK integration files which we can do by copying the folder 'Adsense for Games' from the [Flappy Bird project][flappy-bird-ads-demo].
````
```suggestion
The first step is to add the Google SDK integration files which we can do by copying the folder 'Adsense for Games' from the [Flappy Bird project][flappy-bird-ads-demo].
```

willeastcott commented on content/en/tutorials/google-ads-for-games.md:
```diff
@@ -0,0 +1,377 @@
+    var _beforeAutoRender = null;
+
+    afg.onBeforeAd = function () {
+        // Currently using private API for this. May need to be updated
+        // if PlayCanvas engine is updated
+        _mouseTargetElement = app.mouse._target;
+        _keyboardTargetElement = app.keyboard._element;
+
+        if (app.touch) {
+            _touchTargetElement = app.touch._element;
+        }
```
Not ideal so maybe we can revisit this at some point to avoid messing with engine internals.

yaustar replied:
Yep, it's a bit odd that we have public API that reassigns the target HTML element but not to get the element from the input handlers

willeastcott commented on content/en/tutorials/google-ads-for-games.md:
```diff
@@ -0,0 +1,377 @@
+- `type` which has to be one on these types from the Google SDK documentation
+- `name` which can be any name we give and is used for reporting in AdSense
+- `beforeAd` which is the callback to execute code before the ad is shown
+- `adBreakDone` which is the callback that is called after the ads is shown if even if the ad didn't show for any reason
```
```suggestion
- `adBreakDone` which is the callback that is called after the ad is shown (even if the ad didn't show for any reason)
```

willeastcott commented on content/en/tutorials/google-ads-for-games.md:
```diff
@@ -0,0 +1,377 @@
+Here, we check if the Google SDK is ready first before calling `adBreak` which is the function exposed by the Google SDK itself. More documentation and options can be found on [their site][adbreak-docs]. For the example, we will use:
+
+- `type` which has to be one on these types from the Google SDK documentation
+- `name` which can be any name we give and is used for reporting in AdSense
```
```suggestion
- `name` which can be any name we want and is used for reporting in AdSense
```

willeastcott commented on content/en/tutorials/google-ads-for-games.md:
```diff
@@ -0,0 +1,377 @@
+- `beforeAd` which is the callback to execute code before the ad is shown
+- `adBreakDone` which is the callback that is called after the ads is shown if even if the ad didn't show for any reason
+
+As you can see we use our helper functions `onBeforeAd` and `onAfterAd` in the callbacks to handle the common areas to 'pause' the game before the ad 'resume' afterwards as covered before. You may want to add extra custom logic here for your own game if there are specific areas in your game that you would want to handle.
```
```suggestion
As you can see, we use our helper functions `onBeforeAd` and `onAfterAd` in the callbacks to handle the common areas to 'pause' the game before the ad 'resume' afterwards as covered before. You may want to add extra custom logic here for your own game if there are specific areas in your game that you would want to handle.
```

willeastcott commented on content/en/tutorials/google-ads-for-games.md:
```diff
@@ -0,0 +1,377 @@
+
+As you can see we use our helper functions `onBeforeAd` and `onAfterAd` in the callbacks to handle the common areas to 'pause' the game before the ad 'resume' afterwards as covered before. You may want to add extra custom logic here for your own game if there are specific areas in your game that you would want to handle.
+
+Now lets see it in action
```
```suggestion
Now lets see it in action!
```

willeastcott commented:
This is really cool. Just one thing though. Maybe we want to consider a Monetization section in the User Manual. This still feels like a tutorial though. But yeah, it would be cool to cover the various way devs can monetize their PlayCanvas content more generally.

yaustar commented:
> This is really cool. Just one thing though. Maybe we want to consider a Monetization section in the User Manual. This still feels like a tutorial though. But yeah, it would be cool to cover the various way devs can monetize their PlayCanvas content more generally.

Yep, we have another company I'm working with that I hope can also add to the this section of tutorials/information.

yaustar commented:
Updated title and references for Google Adsense/Ads to Google H5 Games Ads based on early feedback from their team.

Also added steps to sign up for the ad service.

I will update the GIFs and video once I've received full feedback from their team

nbeloglazov commented on content/en/tutorials/google-ads-for-games.md:
````diff
@@ -0,0 +1,388 @@
+
+```
+// https://developers.google.com/ad-placement/docs/example
+(function() {
+    var script = document.createElement('script');
+    script.setAttribute('data-ad-client', 'ca-pub-XXXXXXXXX');
+
+    // Comment out to disable test mode
+    script.setAttribute('data-adbreak-test', 'on');
+
+    // Ad frequency is user configuble
+    // https://developers.google.com/ad-placement/docs/ad-rate
+    script.setAttribute('data-ad-frequency-hint', '30s');
+
+    // Comment in if you would like to use a specific channel for ad reporting
+    // https://developers.google.com/ad-placement/docs/advanced-reporting
+    // script.setAttribute('data-ad-channel', 'XXXXXX');
+
+    script.src = 'https://pagead2.googlesyndication.com/pagead/js/adsbygoogle.js';
+    document.head.appendChild(script);
````
Please update this section. We know use slightly different flavor of adding adsbygoogle.js. It's been used by the AdSense for a while (6+ months) but only know we updated ad-placement docs to reflect it. You can see it here: https://developers.google.com/ad-placement/docs/example#indexhtml_web

The main benefit is that client is now added in adsbygoogle.js url itself so we can faster load ad-placement API. Which is important for preroll ads.

yaustar replied:
Sorry, what would you need it changed to? We can't embed script tags in the html due to the way the Editor projects work if that's what is being suggested?

This function is called first thing in the project, before the engine load and even before the assets are loaded.

yaustar replied:
Oh wait, is it just the url and format that needs to be changed here?

nbeloglazov replied:
Yes. I meant you should add `adsbygoogle.js?client=ca-pub-XXXXXXXXX`, remove `data-ad-client` attribute and add `crossorigin=anonymous` attribute on the constructed script tag.

yaustar commented:
@nbeloglazov Thanks for the feedback, I've updated the scripts in the PlayCanvas projects and also the tutorial text for the change in adding the Ads SDK

Please let me know if this is the only change needed or if it is still in review as I still need to update the GIFs and ideally would like to do them last 😅

nbeloglazov commented on content/en/tutorials/google-ads-for-games.md:
```diff
@@ -0,0 +1,388 @@
+    // https://developers.google.com/ad-placement/docs/advanced-reporting
+    // script.setAttribute('data-ad-channel', 'XXXXXX');
+
+    script.src = 'https://pagead2.googlesyndication.com/pagead/js/adsbygoogle.js?client=ca-pub-2682356268914970';
```
nit: use `ca-pub-XXXXXXXXX` here to match the wording before this code snippet.

yaustar replied:
Thanks!

yaustar commented on content/en/tutorials/google-ads-for-games.md:
```diff
@@ -0,0 +1,388 @@
+    // https://developers.google.com/ad-placement/docs/advanced-reporting
+    // script.setAttribute('data-ad-channel', 'XXXXXX');
+
+    script.src = 'https://pagead2.googlesyndication.com/pagead/js/adsbygoogle.js?client=ca-pub-2682356268914970';
```
```suggestion
    script.src = 'https://pagead2.googlesyndication.com/pagead/js/adsbygoogle.js?client=ca-pub-XXXXXXXXX';
```

nbeloglazov commented:
High-level comment. You use `function() {}` over arrow functions `() => {}` for most of the code. Only adsense-related calls use arrow functions. Do you have a style guide around that? Using just one or another will make overall tutorial more consistent. At google we use arrow functions for most cases as they preserve `this` context and don't need to explicitly bind or pass `this`.

yaustar commented:
> High-level comment. You use `function() {}` over arrow functions `() => {}` for most of the code. Only adsense-related calls use arrow functions. Do you have a style guide around that? Using just one or another will make overall tutorial more consistent. At google we use arrow functions for most cases as they preserve `this` context and don't need to explicitly bind or pass `this`.
> 
> Otherwise looks good.

Good callout, we use ES5 JS for tutorials for PlayCanvas tutorials because our publish pipeline doesn't transpile down yet.

yaustar commented:
Removed YouTube video as it's now outdated and feels like doing it last makes it more difficult to work with. Will need to revise the tutorial/documentation workflow.

Updated GIFs to use the new Google Product name of Google H5 Ads for Games

Updated code to use ES5 style callbacks consistently
